sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,448 @@
+      public Function<Object, Object> visit(ArrowType.Map type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Duration type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Int type) {
+        return null;
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.FloatingPoint type) {
+        return null;
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Utf8 type) {
+        return (Object text) -> ((Text) text).toString();
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Binary type) {
+        return null;
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.FixedSizeBinary type) {
+        return null;
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Bool type) {
+        return null;
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Decimal type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Date type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Time type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Timestamp type) {
+        DateTimeZone tz = DateTimeZone.forID(type.getTimezone());
+        switch (type.getUnit()) {
+          case MICROSECOND:
+            return (epochMicros) -> new DateTime((long) epochMicros / 1000, tz);
+          case MILLISECOND:
+            return (epochMills) -> new DateTime((long) epochMills, tz);
+          default:
+            throw new AssertionError("Encountered unrecognized TimeUnit: " + type.getUnit());
+        }
+      }
+
+      @Override
+      public Function<Object, Object> visit(ArrowType.Interval type) {
+        throw new IllegalArgumentException("Type \'" + type.toString() + "\' not supported.");
+      }
+    }
+
+    public VectorSchemaRootRowIterator(Schema schema, VectorSchemaRoot vectorSchemaRoot) {
+      this.schema = schema;
+      this.vectorSchemaRoot = vectorSchemaRoot;
+      this.fieldValueGetters =
+          new CachingFactory<>(
+              FieldVectorListValueGetterFactory.of(vectorSchemaRoot.getFieldVectors()));
+      this.currRowIndex = 0;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return currRowIndex < vectorSchemaRoot.getRowCount();
+    }
+
+    @Override
+    public Row next() {
+      Row result =
```

Comment:
should there be a precondition check here that to assert there are more rows?

Reply:
yes let's add this